Networking/Networking/CookieNonce/CookieNonceAuthenticator.swift

```diff
@@ -0,0 +1,169 @@
+            }
+        }
+    }
+
+    func handleSiteCredentialLogin(manager: SessionManager) async throws {
+        let request = authenticatedRequest()
+        return try await withCheckedThrowingContinuation { continuation in
+            manager.request(request)
+                .validate()
+                .response { response in
+                    if let error = response.error {
+                        continuation.resume(throwing: error)
+                    } else {
+                        continuation.resume(returning: ())
+                    }
+                }
+        }
+    }
+
+    func handleNonceRetrieval(request: URLRequest, manager: SessionManager) async throws -> String {
+        try await withCheckedThrowingContinuation { continuation -> Void in
+            manager.request(request)
+                .validate()
+                .responseString { response in
+                    switch response.result {
+                    case .failure(let error):
+                        continuation.resume(throwing: error)
+                    case .success(let page):
+                        continuation.resume(returning: page)
+                    }
+                }
+        }
+    }
+
+    func successfulLoginSequence() {
+        DDLogInfo("Completed Cookie+Nonce login sequence for \(loginURL)")
+        completeRequests(true)
+    }
+
+    func invalidateLoginSequence(error: Error) {
+        canRetry = false
+        if case .postLoginFailed(let originalError) = error {
+            let nsError = originalError as NSError
+            if nsError.domain == NSURLErrorDomain, nsError.code == NSURLErrorNotConnectedToInternet {
+                canRetry = true
+            }
+        }
+        DDLogInfo("Aborting Cookie+Nonce login sequence for \(loginURL)")
+        completeRequests(false)
+        isAuthenticating = false
+    }
+
+    func completeRequests(_ shouldRetry: Bool) {
+        requestsToRetry.forEach { (completion) in
+            completion(shouldRetry, 0.0)
+        }
+        requestsToRetry.removeAll()
+    }
+
+    func authenticatedRequest() -> URLRequest {
+        var request = URLRequest(url: loginURL)
+
+        request.httpMethod = "POST"
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+
+        var parameters = [URLQueryItem]()
+        parameters.append(URLQueryItem(name: "log", value: username))
+        parameters.append(URLQueryItem(name: "pwd", value: password))
+        parameters.append(URLQueryItem(name: "rememberme", value: "true"))
+        var components = URLComponents()
+        components.queryItems = parameters
+        request.httpBody = components.percentEncodedQuery?.data(using: .utf8)
+        return request
+    }
+
+    func readNonceFromAjaxAction(html: String) -> String? {
+        guard !html.isEmpty else {
+            return nil
+        }
+        return html
```

Comment:
nit: maybe can simplify?
```suggestion
        html.isEmpty ? nil: html
```